# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: docshell/shistory/PSHistory.ipdl
# Commit: 6ebc7618a5fa
# Full Hash: 6ebc7618a5fadb59585a4d8b9f17eb65e5654cc6
# Author: Anny Gakhokidze <agakhokidze@mozilla.com>
# Date: 2020-03-10 21:44:54
# Regressor Bug: 1546759
# File Overlap Count: 2
# Description:
#   Bug 1546759 - Change nsSHistory::WalkHistoryEntries to walk browsing context tree instead of doc shell tree, r=peterv,nika
#   
#   WalkHistoryEntries function gets called by nsSHistory::CloneAndReplaceChild
#   and nsSHistory::SetChildHistoryEntry recursively, so those have to be moved
#   into the parent process. This eliminates many sync IPC calls.
# ==============================================================================

diff -r ed75364b23c3 -r 6ebc7618a5fa docshell/shistory/PSHistory.ipdl
--- a/docshell/shistory/PSHistory.ipdl	Tue Mar 10 11:39:06 2020 +0000
+++ b/docshell/shistory/PSHistory.ipdl	Tue Mar 10 14:28:22 2020 +0000
@@ -62,6 +62,22 @@
   async EnsureCorrectEntryAtCurrIndex(PSHEntry entry);
   async EvictContentViewersOrReplaceEntry(nullable PSHEntry newSHEntry, bool replace);
   async NotifyListenersContentViewerEvicted(uint32_t numEvicted);
+  // See below for some explanation
+  sync AddToRootSessionHistory(bool cloneChildren, nullable PSHEntry OSHE, MaybeDiscardedBrowsingContext BC, PSHEntry entry,
+                                uint32_t loadType, bool shouldPersist) returns (int32_t? previousEntryIndex, int32_t? loadedEntryIndex,
+                                SwapEntriesDocshellData[] entriesToUpdate, int32_t entriesPurged, nsresult result);
+  // In CloneAndReplaceChild and SetChildHistoryEntry methods (which get called within this
+  // and above function) there are calls to SwapHistoryEntries to update entries in the docshell.
+  // We don't have access to the docshell in the parent
+  // process so we have to either wait to update the docshell or update it via an IPC call.
+  // We pass in the ChildID of the process who invoked the IPC to make decisions
+  // about whether we should update entries in the docshell via an IPC call -
+  // if the caller is a process different from where the docshell lives.
+  // Else, we return entries that we need to update in entriesToUpdate array
+  // and update the docshell with such entries in the child process.
+  sync AddChildSHEntryHelper(PSHEntry cloneRef, PSHEntry newEntry, MaybeDiscardedBrowsingContext BC,
+                             bool cloneChildren) returns (SwapEntriesDocshellData[] entriesToUpdate,
+                             int32_t entriesPurged, CrossProcessSHEntry entry, nsresult result);
   async __delete__();
 
 child: